core/base/mpi.cpp

```diff
@@ -0,0 +1,101 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2022, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/mpi.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+
+#include <mpi.h>
+#include <string>
+
+
+namespace gko {
+namespace mpi {
+
+
+int map_rank_to_device_id(MPI_Comm comm, const int num_devices)
+{
+    GKO_ASSERT(num_devices > 0);
+    if (num_devices == 1) {
+        return 0;
+    } else {
+        auto mpi_node_local_rank = [](MPI_Comm comm_) {
+            int local_rank;
+            MPI_Comm local_comm;
+            GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_split_type(
+                comm_, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &local_comm));
+            GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_rank(local_comm, &local_rank));
+            MPI_Comm_free(&local_comm);
+            return local_rank;
+        };
+
+        // When we are using MPI_COMM_WORLD, there might be already an
+        // environment variable describing the node local rank, so we
+        // prioritize it. If no suitable environment variable is found
+        // we determine the node-local rank with MPI calls.
+        int local_rank;
+        int compare_result;
+        int is_initialized;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Initialized(&is_initialized));
+        if (is_initialized) {
+            GKO_ASSERT_NO_MPI_ERRORS(
+                MPI_Comm_compare(comm, MPI_COMM_WORLD, &compare_result));
```

Comment:
why does this function need to work at all if MPI is not yet initialized? I would actually prefer if it could fail loudly in that case

Reply:
If the id was only deduced from the env variables, it could work without MPI initialized. But I think it's better to also have the comm-splitting approach available. This can be used if ginkgo should not use MPI_COMM_WORLD, but a different communicator, so I would like to keep that approach.

I think if I remove the checks, then it will throw if MPI is not initialized, which should be loud enough.